src/backend/distributed/connection/connection_management.c

```diff
@@ -454,13 +467,97 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 			continue;
 		}
 
+		if ((flags & REQUIRE_METADATA_CONNECTION) &&
+			!connection->useForMetadataOperations)
+		{
+			/*
+			 * The caller requested a metadata connection, and this is not the
+			 * metadata connection.
+			 */
+			continue;
```

Comment:
what if we don't have a metadata connection yet? can we use an existing one?

Reply:
we handle that below, after the loop? We pick the first connection as the metadata connection, and mark `useForMetadataOperations`